File: blockchain/src/nodes/nodes_service.py
import requests, json

from rest_framework.response import Response
from rest_framework import status

from adapters.factory import DjangoStorageFactory
from blockchain.libraries.factory import LibraryFactory
from blockchain.src.registry.registry_service import RegistryService
from blockchain.libraries.block import Block

class NodeService():
    
    def __init__(self, storage: DjangoStorageFactory, library: LibraryFactory):
        self.storage = storage
        self.library = library

    def new_node(self, payload: dict):
        addr = payload.get("node_address")

        if not addr:
            return Response({"message": "Missing node_address field!"}, status.HTTP_400_BAD_REQUEST)
        
        #TODO: check if node exist before insert and communicate new nodes to peers
        self.storage.createPeersModel().insert({ "ip_address": addr })
        return RegistryService(self.storage, self.library).list()

    def join_network(self, payload: dict, host: str):
        addr = payload.get("node_address")
        if not addr:
            return Response({"message": "Missing node_address field!"}, status.HTTP_400_BAD_REQUEST)

        data = {"node_address": host}
        headers = {'Content-Type': "application/json"}

        # Make a request to register with remote node and obtain information
        response = requests.post(f"http://{addr}/node/register", data=json.dumps(data), headers=headers)
        if response.status_code == 200:
            payload = response.json()
            # update chain and the peers
            self.library.createBlockchain().create_chain_from_dump(payload['chain'])
            # #nesse nó ou em um novo nó DESENHAR!
            self.library.createPeersManager().sync_peers([addr, *payload['peers']], host)

            return Response({"message":"Registration successful"}, status.HTTP_200_OK)
        else:
            return Response({"message": "Error registering node in network."}, status.HTTP_500_INTERNAL_SERVER_ERROR)

    def sync_block(self, block: dict):
        proof = block['hash']
        for key in ['id', 'hash', 'created_at']:
            if key in block: del block[key]

        block = Block(**block)
        added = self.library.createBlockchain().add_block(block, proof)

        if not added: return Response({"message":"The block is discarded by the node."}, status.HTTP_500_INTERNAL_SERVER_ERROR)

        return Response({"message": "Block added to the chain"}, status.HTTP_201_CREATED)

    #TODO: remove method after tests
    def clear_local(self):
        self.storage.createPeersModel().delete()
        self.storage.createBlockModels().delete()
        return  Response({"message": "Clear complete!"}, status.HTTP_200_OK)
        

File: blockchain/src/nodes/nodes_service.ts
import requests
import json
import { Response, HttpStatus } from '@nestjs/common';

import { DjangoStorageFactory } from '../adapters/factory/django-storage.factory';
import { LibraryFactory } from '../blockchain/libraries/factory/library.factory';
import { RegistryService } from '../blockchain/src/registry/registry.service';
import { Block } from '../blockchain/libraries/block/block';

// Classe que lida com comunicação e armazenamento de dados relacionados à blockchain
export class NodeService {

    private storage: DjangoStorageFactory;
    private library: LibraryFactory;

    constructor(storage: DjangoStorageFactory, library: LibraryFactory) {
        this.storage = storage;
        this.library = library;
    }

    /*
    Verifica se o endereço do novo nó está presente e, em caso afirmativo, adiciona o novo nó ao armazenamento de pares (peers) e retorna a lista de pares atualizada
    */
    public async newNode(payload: { node_address: string }): Promise<Response> {
        const addr = payload.node_address;

        if (!addr) {
            return { message: 'Missing node_address field!' }, HttpStatus.BAD_REQUEST;
        }

        this.storage.createPeersModel().insert({ ip_address: addr });
        return new RegistryService(this.storage, this.library).list();
    }

    /*
    Verifica se o endereço do novo nó está presente e, em caso afirmativo, envia uma solicitação POST para o endereço especificado para registrar o novo nó
    */
    public async joinNetwork(payload: { node_address: string }, host: string): Promise<Response> {
        const addr = payload.node_address;

        if (!addr) {
            return { message: 'Missing node_address field!' }, HttpStatus.BAD_REQUEST;
        }

        const data = { node_address: host };
        const headers = { 'Content-Type': 'application/json' };

        const response = await requests.post(
            `http://${addr}/node/register`, { data, headers });

        if (response.statusCode === 200) {
            const responsePayload = response.json();
            this.library.createBlockchain().createChainFromDump(responsePayload.chain);
            this.library.createPeersManager().syncPeers([addr, ...responsePayload.peers], host);
            return { message: 'Registration successful' }, HttpStatus.OK;
        } else {
            return { message: 'Error registering node in network.' }, HttpStatus.INTERNAL_SERVER_ERROR;
        }
    }

    /*
    Tenta adicionar um bloco à cadeia de blocos
    */
    public async syncBlock(block: Block): Promise<Response> {
        const proof = block.hash;
        delete block.id;
        delete block.hash;
        delete block.createdAt;

        const added = this.library.createBlockchain().addBlock(block, proof);

        if (!added) {
            return { message: 'The block is discarded by the node.' }, HttpStatus.INTERNAL_SERVER_ERROR;
        }

        return { message: 'Block added to the chain' }, HttpStatus.CREATED;
    }

    /*
     Limpa o armazenamento local de pares e blocos
    */
    public async clearLocal(): Promise<Response> {
        this.storage.createPeersModel().delete();
        this.storage.createBlockModels().delete();
        return { message: 'Clear complete!' }, HttpStatus.OK;
    }
}
